### ISSUE TITLE
Split MediaQuery into ViewQuery and PlatformQuery

### ISSUE THREAD

**Issue by goderbauer** [flutter/flutter#115375]
MediaQuery mixes platform-specific concepts (that are shared among all views) with view-specific concepts. This makes MediaQuery cumbersome to use in a multi view environment: The View widget needs to inject the view-specific data into the widget tree, without overriding the platform-specific data from the surrounding context. This can be done with the existing MediaQuery concept, but it would require copying the platform-specific data from the context and makes rebuilds less targeted when the data changes. 

Since we are removing single view assumptions elsewhere in the framework as well, it would be cleaner to split MediaQuery into ViewQuery (which contains the view-specific data) and PlatformQuery (which contains the rest). The data from the PlatformQuery would be injected near the root of the tree and is accessible from all views. The data from the ViewQuery would be injected by the View widget and its only available within that View.

If we implement the InheritanceBoundary (https://github.com/flutter/flutter/issues/115381), the PlatformQuery would not be an inheritance boundary, while ViewQuery would be one.

Data moving to ViewQuery:
* devicePixelRatio
* displayFeatures
* orientation
* padding
* size
* systemGestureInsets
* viewInsets
* viewPadding

Data moving to PlatformQuery:
* accessibleNavigation
* alwaysUse24HourFormat 
* boldText
* disableAnimations
* gestureSettings
* highContrast
* invertColors
* navigationMode
* platformBrightness
* textScaleFactor

Migration: We would deprecate MediaQuery and redirect people to PlatformQuery and ViewQuery. During the deprecation policy, MediaQuery would be reimplemented by composing PlatformQuery and ViewQuery internally.

**Comment by goderbauer** [comment]
cc @gspencergoog @Hixie @LongCatIsLooong @a-wallen @dkwingsmt @loic-sharma for thoughts

**Comment by goderbauer** [comment]
ViewQuery and PlatformQuery should be implemented as an InheritedModel, to allow developers to only depend on a certain aspect (similar to the refactoring done in https://github.com/flutter/flutter/pull/114459 for MediaQuery).

Furthermore, PlatformQuery could also inject the PlatformDispatcher into the tree and ViewQuery the FlutterView.

**Comment by Hixie** [comment]
I think we should probably only do this if we can do the migration automatically and make sure that migration keeps working for years after we remove `MediaQuery`. This is such a core widget that it would likely break literally every application ever.

**Comment by goderbauer** [comment]
After talking to some of the dartfix experts I don't think there is a way to auto-migrate people from something like

```dart
MediaQuery.of(context);
MediaQuery.of(context).textScaleFactor;
MediaQuery.of(context).size;
```

to 

```dart
PlatformQuery.of(context); // or ViewQuery.of(context);
PlatformQuery.of(context).textScaleFactor;
ViewQuery.of(context).size;
```

Dartfix isn't able to look at the context to decide whether `MediaQuery.of(context)` needs to turn into `PlatformQuery.of(context)` or `ViewQuery.of(context)`. Sometimes, the context to decide is very close by (the property accessed on the return value, see examples 2 and 3 above). But in other cases, the context could be very far away (case 1 above) - and could potentially require that you actually need to obtain both, the PlatformQuery and the ViewQuery, which makes the migration even more complicated.

The MediaQuery API is also very popular, google3 alone lists over 1500 calls to `MediaQuery.of`.

Long story short, splitting MediaQuery into ViewQuery and PlatformQuery is a non-started because the MediaQuery is too popular and we can't offer an auto-migration path.

**Comment by goderbauer** [comment]
Alternative Idee:

The `View` widget will just call `MediaQuery.of(context)` to obtain the surrounding `MediaQueryData` and then call `MediaQueryData.copyWith` to replace the view-specific data in the surrounding MediaQuery with the data from the `FlutterView` associated with the `View` widget. The new `MediaQueryData` is injected into the tree for descendants of the `View`. 

If there is no surrounding `MediaQueryData`, the `View` widget instantiates a new one and fills in the platform-specific data from the `PlatformDispatcher` of the `FlutterView` associated with the `View` widget.

In this world, the `View` widget is the one responsible for injecting the MediaQuery into the widget tree. Outside of a `View`, you wouldn't be able to access the `MediaQueryData`. In theory, we could create a new MediaQueryData constructor in the future that would create a `MediaQueryData` object containing only platform-specific data (accessing any of the view-specific data would throw). That way, the platform-specific data of `MediaQuery` could be made available in parts of the widget tree that is not part of a `View`. We would probably also need a way to check whether view-specific data is available on the MediaQueryData in this world, which would make usage and API rather complicated. As long as we don't have a use case for this, we probably shouldn't make the MediaQueryData available outside of `View`s.

**Comment by dkwingsmt** [comment]
> Alternative Idee:
> 
> The `View` widget will just call `MediaQuery.of(context)` to obtain the surrounding `MediaQueryData` and then call `MediaQueryData.copyWith` to replace the view-specific data in the surrounding MediaQuery with the data from the `FlutterView` associated with the `View` widget. The new `MediaQueryData` is injected into the tree for descendants of the `View`.

If I understand correctly, this forbids developers to define inherited widget classes that also need their inherited data partially cleared on view boundary. Not perfect either. :(

**Comment by moffatman** [comment]
Any more possibilities here, or are you sure it's not feasible? If so we can move forward on https://github.com/flutter/flutter/pull/114459?
